File: tests/sdk/nodejs/test/apigateway-websocket.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateApiCommand,
  CreateIntegrationCommand,
  CreateRouteCommand,
  CreateStageCommand,
  DeleteApiCommand,
} from "@aws-sdk/client-apigatewayv2";
import {
  CreateRoleCommand,
  DeleteRoleCommand,
} from "@aws-sdk/client-iam";
import {
  AddPermissionCommand,
  CreateFunctionCommand,
  DeleteFunctionCommand,
} from "@aws-sdk/client-lambda";
import {
  endpoint,
  newAPIGWv2,
  newIAM,
  newLambda,
  requireReady,
  uniquePrefix,
} from "../lib/helpers.mjs";

const TRUST =
  '{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"},"Action":"sts:AssumeRole"}]}';

function crc32(buf) {
  let c = 0xffffffff;
  for (let i = 0; i < buf.length; i++) {
    c ^= buf[i];
    for (let k = 0; k < 8; k++) {
      c = c & 1 ? (c >>> 1) ^ 0xedb88320 : c >>> 1;
    }
  }
  return (c ^ 0xffffffff) >>> 0;
}

function minimalPythonZip() {
  const payload = Buffer.from(
    "def handler(event, context):\n    return {'ok': True}\n",
    "utf8",
  );
  const name = Buffer.from("index.py");
  const size = payload.length;
  const crc = crc32(payload);
  const local = Buffer.alloc(30 + name.length);
  local.writeUInt32LE(0x04034b50, 0);
  local.writeUInt16LE(20, 4);
  local.writeUInt16LE(0, 6);
  local.writeUInt16LE(0, 8);
  local.writeUInt16LE(0, 10);
  local.writeUInt16LE(0, 12);
  local.writeUInt32LE(crc, 14);
  local.writeUInt32LE(size, 18);
  local.writeUInt32LE(size, 22);
  local.writeUInt16LE(name.length, 26);
  local.writeUInt16LE(0, 28);
  name.copy(local, 30);
  const central = Buffer.alloc(46 + name.length);
  central.writeUInt32LE(0x02014b50, 0);
  central.writeUInt16LE(20, 4);
  central.writeUInt16LE(20, 6);
  central.writeUInt16LE(0, 8);
  central.writeUInt16LE(0, 10);
  central.writeUInt16LE(0, 12);
  central.writeUInt16LE(0, 14);
  central.writeUInt32LE(crc, 16);
  central.writeUInt32LE(size, 20);
  central.writeUInt32LE(size, 24);
  central.writeUInt16LE(name.length, 28);
  central.writeUInt16LE(0, 30);
  central.writeUInt16LE(0, 32);
  central.writeUInt16LE(0, 34);
  central.writeUInt16LE(0, 36);
  central.writeUInt32LE(0, 38);
  central.writeUInt32LE(0, 42);
  name.copy(central, 46);
  const end = Buffer.alloc(22);
  end.writeUInt32LE(0x06054b50, 0);
  end.writeUInt16LE(0, 4);
  end.writeUInt16LE(0, 6);
  end.writeUInt16LE(1, 8);
  end.writeUInt16LE(1, 10);
  end.writeUInt32LE(central.length, 12);
  end.writeUInt32LE(local.length + size, 16);
  end.writeUInt16LE(0, 20);
  return Buffer.concat([local, payload, central, end]);
}

test("WebSocket API create routes soft-skip connect without nested", async (t) => {
  await requireReady();
  const prefix = uniquePrefix();
  const iam = newIAM();
  const lam = newLambda();
  const apigw = newAPIGWv2();

  const roleName = `${prefix}-ws-role`;
  const role = await iam.send(
    new CreateRoleCommand({
      RoleName: roleName,
      AssumeRolePolicyDocument: TRUST,
    }),
  );
  t.after(async () => {
    await iam.send(new DeleteRoleCommand({ RoleName: roleName })).catch(() => {});
  });
  const roleArn = role.Role.Arn;

  const fnName = `${prefix}-ws-fn`;
  const fn = await lam.send(
    new CreateFunctionCommand({
      FunctionName: fnName,
      Runtime: "python3.12",
      Role: roleArn,
      Handler: "index.handler",
      Code: { ZipFile: minimalPythonZip() },
    }),
  );
  t.after(async () => {
    await lam
      .send(new DeleteFunctionCommand({ FunctionName: fnName }))
      .catch(() => {});
  });
  await lam.send(
    new AddPermissionCommand({
      FunctionName: fnName,
      StatementId: `${prefix}-ws-perm`,
      Action: "lambda:InvokeFunction",
      Principal: "apigateway.amazonaws.com",
    }),
  );

  const api = await apigw.send(
    new CreateApiCommand({
      Name: `${prefix}-ws-api`,
      ProtocolType: "WEBSOCKET",
    }),
  );
  assert.equal(api.ProtocolType, "WEBSOCKET");
  const apiId = api.ApiId;
  t.after(async () => {
    await apigw.send(new DeleteApiCommand({ ApiId: apiId })).catch(() => {});
  });

  const integ = await apigw.send(
    new CreateIntegrationCommand({
      ApiId: apiId,
      IntegrationType: "AWS_PROXY",
      IntegrationUri: fn.FunctionArn,
    }),
  );
  for (const routeKey of ["$connect", "$disconnect", "$default"]) {
    await apigw.send(
      new CreateRouteCommand({
        ApiId: apiId,
        RouteKey: routeKey,
        Target: `integrations/${integ.IntegrationId}`,
        AuthorizationType: "NONE",
      }),
    );
  }
  await apigw.send(
    new CreateStageCommand({
      ApiId: apiId,
      StageName: "$default",
      AutoDeploy: true,
    }),
  );

  const res = await fetch(
    `${endpoint()}/ws-api/${apiId}/$default/$connect`,
    { method: "POST" },
  );
  const text = await res.text();
  if (
    res.status === 503 &&
    (text.includes("compute unavailable") ||
      process.env.NOCTAXRIS_NESTED !== "1")
  ) {
    t.skip(
      "WebSocket $connect soft-skip: nested Lambda unavailable (set NOCTAXRIS_NESTED=1)",
    );
    return;
  }
  assert.equal(res.status, 200, text);
  const body = JSON.parse(text);
  assert.ok(body.connectionId);
});

File: tests/sdk/go/apigateway_websocket_test.go
package sdk_test

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/apigatewayv2"
	apitypes "github.com/aws/aws-sdk-go-v2/service/apigatewayv2/types"
	"github.com/aws/aws-sdk-go-v2/service/iam"
	"github.com/aws/aws-sdk-go-v2/service/lambda"
	"github.com/aws/aws-sdk-go-v2/service/lambda/types"
)

func TestWebSocketAPICreateRoutesAndSoftSkipConnect(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	iamClient := newIAM(t, cfg)
	lam := newLambda(t, cfg)
	apigw := newAPIGWv2(t, cfg)
	ctx := context.Background()
	prefix := uniquePrefix(t)
	httpClient := &http.Client{Timeout: 30 * time.Second}

	roleName := prefix + "-ws-role"
	trust := `{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"},"Action":"sts:AssumeRole"}]}`
	roleOut, err := iamClient.CreateRole(ctx, &iam.CreateRoleInput{
		RoleName:                 aws.String(roleName),
		AssumeRolePolicyDocument: aws.String(trust),
	})
	if err != nil {
		t.Fatalf("CreateRole: %v", err)
	}
	roleARN := *roleOut.Role.Arn
	t.Cleanup(func() {
		_, _ = iamClient.DeleteRole(context.Background(), &iam.DeleteRoleInput{RoleName: aws.String(roleName)})
	})

	fnName := prefix + "-ws-fn"
	fnOut, err := lam.CreateFunction(ctx, &lambda.CreateFunctionInput{
		FunctionName: aws.String(fnName),
		Runtime:      types.RuntimePython312,
		Role:         aws.String(roleARN),
		Handler:      aws.String("index.handler"),
		Code:         &types.FunctionCode{ZipFile: minimalPythonZip(t)},
	})
	if err != nil {
		t.Fatalf("CreateFunction: %v", err)
	}
	lambdaARN := *fnOut.FunctionArn
	t.Cleanup(func() {
		_, _ = lam.DeleteFunction(context.Background(), &lambda.DeleteFunctionInput{FunctionName: aws.String(fnName)})
	})
	_, err = lam.AddPermission(ctx, &lambda.AddPermissionInput{
		FunctionName: aws.String(fnName),
		StatementId:  aws.String(prefix + "-ws-perm"),
		Action:       aws.String("lambda:InvokeFunction"),
		Principal:    aws.String("apigateway.amazonaws.com"),
	})
	if err != nil {
		t.Fatalf("AddPermission: %v", err)
	}

	apiOut, err := apigw.CreateApi(ctx, &apigatewayv2.CreateApiInput{
		Name:         aws.String(prefix + "-ws-api"),
		ProtocolType: apitypes.ProtocolTypeWebsocket,
	})
	if err != nil {
		t.Fatalf("CreateApi WEBSOCKET: %v", err)
	}
	apiID := *apiOut.ApiId
	t.Cleanup(func() {
		_, _ = apigw.DeleteApi(context.Background(), &apigatewayv2.DeleteApiInput{ApiId: aws.String(apiID)})
	})
	if apiOut.ProtocolType != apitypes.ProtocolTypeWebsocket {
		t.Fatalf("ProtocolType=%v", apiOut.ProtocolType)
	}

	intOut, err := apigw.CreateIntegration(ctx, &apigatewayv2.CreateIntegrationInput{
		ApiId:           aws.String(apiID),
		IntegrationType: apitypes.IntegrationTypeAwsProxy,
		IntegrationUri:  aws.String(lambdaARN),
	})
	if err != nil {
		t.Fatalf("CreateIntegration: %v", err)
	}
	integrationID := *intOut.IntegrationId
	for _, rk := range []string{"$connect", "$disconnect", "$default"} {
		_, err := apigw.CreateRoute(ctx, &apigatewayv2.CreateRouteInput{
			ApiId:             aws.String(apiID),
			RouteKey:          aws.String(rk),
			Target:            aws.String("integrations/" + integrationID),
			AuthorizationType: apitypes.AuthorizationTypeNone,
		})
		if err != nil {
			t.Fatalf("CreateRoute %s: %v", rk, err)
		}
	}
	_, err = apigw.CreateStage(ctx, &apigatewayv2.CreateStageInput{
		ApiId:      aws.String(apiID),
		StageName:  aws.String("$default"),
		AutoDeploy: aws.Bool(true),
	})
	if err != nil {
		t.Fatalf("CreateStage: %v", err)
	}

	req, err := http.NewRequest(http.MethodPost, endpoint()+"/ws-api/"+apiID+"/$default/$connect", nil)
	if err != nil {
		t.Fatal(err)
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		t.Fatalf("lab $connect: %v", err)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode == http.StatusServiceUnavailable && strings.Contains(string(body), "compute unavailable") {
		t.Skip("WebSocket $connect soft-skip: nested Lambda unavailable (set NOCTAXRIS_NESTED=1)")
	}
	if os.Getenv("NOCTAXRIS_NESTED") != "1" && resp.StatusCode == http.StatusServiceUnavailable {
		t.Skip("WebSocket $connect soft-skip: compute unavailable without nested engine")
	}
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("$connect status=%d body=%s", resp.StatusCode, body)
	}
	var connectResp map[string]any
	if err := json.Unmarshal(body, &connectResp); err != nil {
		t.Fatalf("connect json: %v body=%s", err, body)
	}
	if connectResp["connectionId"] == nil || connectResp["connectionId"] == "" {
		t.Fatalf("missing connectionId: %v", connectResp)
	}
}
